### PR TITLE
Updating list-ops test generic types parameters

### PR DESCRIPTION
**Pull request by Shoghy**
I added and deleted generic type parameters from the list-ops exercise tests.

I did this because I was having problems compiling the code and I also noticed that several generic types were redudant, considering that the `List` class has a generic type like `Array`.

I will list 2 of the changed tests and the reasons why I added or removed the generic type parameter, the other tests were changed for the same reasons:

```ts
/*...*/
xit('list to empty list', () => {
  const list1 = List.create<number>()
  const list2 = List.create(1, 2, 3, 4)
  expect(list1.append(list2)).toEqual(list2)
})
/*...*/
```
In this first test I added the generic type parameter `<number>` to `list1` so that, even though the constant `list1` has no elements, TypeScript understands that it stores the same type value as `list2`. I did not add `<number>` to `list2` because the type of value it stores is automatically deduced by the parameters sent in the `create` function.
___

```ts
/*...*/
describe('filter list returning only values that satisfy the filter function', () => {
  xit('empty list', () => {
    const list1 = List.create<number>()
    expect(list1.filter((el) => el % 2 === 1)).toHaveValues()
  })
/*...*/
```
In this case I understand that `<number>` in `filter` refers to the type of value that the list returned by `filter` will store, however this is not necessary if `List` has a generic type, since the returned list will store the same type of value as the original list.

### CODE REVIEW

**Comment by Shoghy** [comment]
@SleeplessByte

**Comment by SleeplessByte** [comment]
@Shoghy I have rebased this for you. You must also update proof.ci.ts in `.meta`, for the linter to be satisfied. Can you make that change?

**Comment by Shoghy** [comment]
> @Shoghy I have rebased this for you. You must also update `proof.ci.ts` in `.meta`, for the linter to be satisfied. Can you make that change? You'll then want to check the linter.

How do I test that file so I can be sure my changes are all right?

**Comment by SleeplessByte** [comment]
@Shoghy from [CONTRIBUTING.md](https://github.com/exercism/typescript/blob/main/CONTRIBUTING.md)

> If the ASSIGNMENT environment variable is set, only that exercise is tested. For example, if you only want to test the example.js for two-fer, you may, depending on your environment, use:
>
> ```shell
> ASSIGNMENT=practice/two-fer corepack yarn test
> ```

So in your case it would be

```shell
ASSIGNMENT=practice/list-ops corepack yarn test
```

or if you are on windows

```shell
corepack yarn dlx cross-env ASSIGNMENT=practice/two-fer corepack yarn test
```

**Comment by SleeplessByte** [comment]
I'll reformat the file now:

/format

File: exercises/practice/list-ops/.meta/proof.ci.ts [hunk 1]
```diff
@@ -10,17 +10,17 @@ const Null: Cons = {
     return this.value
   },
 
-  push(item): Cons {
+  push<T>(item: T): Cons<T> {
     // eslint-disable-next-line @typescript-eslint/no-use-before-define
-    return new Cons(item, this)
+    return new Cons(item, this) as Cons<T>
```
Comment (SleeplessByte) [inline exercises/practice/list-ops/.meta/proof.ci.ts:15]: Why is this necessary? Doesn't `new Cons(item, this)` already infer `Const<T>`? If not, shouldn't it be `new Cons<T>(item, this)`?

File: exercises/practice/list-ops/.meta/proof.ci.ts [hunk 2]
```diff
@@ -38,27 +38,27 @@ const Null: Cons = {
...
+    return this as Cons<TReturn>
   },
 }
-class Cons {
+class Cons<T> {
   constructor(
-    public readonly value: unknown,
-    public next: Cons = Null
+    public readonly value: T,
+    public next: Cons<T> = Null as Cons<T>
```
Comment (SleeplessByte) [inline exercises/practice/list-ops/.meta/proof.ci.ts:54]: It's probably better to define it as:

```typescript
public next: Cons<T> | Cons<undefined>
```

instead of needing the `as Cons<T>`

File: exercises/practice/list-ops/.meta/proof.ci.ts (hunk 1 above)
Comment (Shoghy) [reply exercises/practice/list-ops/.meta/proof.ci.ts:15]: In this case `Cons` would be of type `Cons<T | undefined>`, however because of how `Null` works the `undefined` part can be omitted, and since is of type `Cons<T | undefined>` using `new Cons<T>(item, this)` throws an error on TypeScript.

File: exercises/practice/list-ops/.meta/proof.ci.ts (hunk 2 above)
Comment (Shoghy) [reply exercises/practice/list-ops/.meta/proof.ci.ts:54]: This can make the code a little bit messier with the types, and the only method where this will be important is `get`, so that's why I made the return value of `get` be `T | undefined`

**Review by SleeplessByte — APPROVED** [review]
I think overall it's a net improvement (less explicit types in the tests), but I am not completely convinced about the proof solution. However let's get this merged and make further improvements in future PRs

Thank you for working on this!
